tests.ml

```diff
@@ -502,13 +457,61 @@ jmp:
 end:
 "
 
-let do_test_dom prog = function () ->
-  let cfg = Cfg.of_program prog in
-  let doms = Cfg.dominators (prog, cfg) in
-  let c = Cfg.common_dominator (test_df2, cfg, doms) [12; 19] in
-  let expected = Cfg.node_at cfg 9 in
-  let open Cfg in
-  assert_equal c.id expected.id
+let do_test_codemotion = function () ->
+  let main p : instruction_stream = List.assoc "main" p in
+  let t = parse_test "
+       goto bla
+      loop:
+       y <- z
+       x <- (x + y)
+       branch (x==10) end loop
+      end:
+       stop
+      bla:
+       const z = 1
+       mut x = 1
+       mut y = z
+       goto loop
+  " in
+  let res = Transform.hoist_assignment t in
+  assert_equal (main res).(8) (Decl_mut ("y0", Some (Simple (Var "z"))));
+  let t = parse_test "
+       mut x = 1
+       mut y = 2
+      loop:
+       y <- 1
+       x <- (x + y)
+       branch (x==10) end loop
+      end:
+  " in
+  let res = Transform.hoist_assignment t in
+  assert_equal (main res).(0) (Decl_mut ("y0", Some (Simple (Lit (Int 1)))));
+  let t = parse_test "
+       mut x = 1
+       mut y = 2
+      loop:
+       y <- x
+       x <- (x + y)
+       branch (x==10) end loop
+      end:
+  " in
+  let res = Transform.hoist_assignment t in
+  assert_equal (main res).(3) (main t).(3);
```

Comment:
It would also be nice to make it more explicit that in this case (and the next one), code motion can't be applied. Maybe in a comment that also explains why?